### PR TITLE
Switch to Angular CLI

### LINKED ISSUE #225: Update project to use Angular CLI
When we initially created `dspace-angular`, the Angular CLI project was very new and unstable.

However, Angular CLI is now the recommended way to create/manage/maintain Angular applications. https://cli.angular.io/

We should update `dspace-angular` to use Angular CLI.  Here's some rough notes on how to update an existing Angular app to use Angular CLI: https://github.com/angular/angular-cli/blob/master/docs/documentation/stories/moving-into-the-cli.md

As an incentive, it was also recently announced that most Angular CLI apps can now be run *directly in your browser* (i.e. no need to even download) using stackblitz.com.  See this Angular.io blog post for more information: https://blog.angular.io/run-angular-cli-repos-directly-in-your-browser-41332fd80901

### PR DESCRIPTION
**Pull request by artlowel**
This PR closes #225 

It turns dspace-angular in to an Angular CLI project.

Key differences are:
- The project now runs on port 4000 for production and port 4200 for development
- Development mode uses watch by default
- The preboot property does not exist anymore, in cli the server side is built for production builds only. As a consequence dev builds are a lot faster.
- The configuration files were moved from ./config/ to ./src/environments. This is where Angular CLI stores their configuration files by default.
- The environment-files are now .ts files instead of .js files, so are the webpack files
- The environment variables are not being injected anymore but simply imported

Known issues:
- Overlaying themes doesn't work yet. Currently only the default theme can be used.  
- Loading sass variables as a typescript module doesn't work yet. The variables we had (e.g. the grid breakpoints) are hardcoded in typescript for now
- There are a number of tests that now throw a warning saying the test "has no expectations". This is usually because those tests contain an asynchronous element, and the test wasn't set up to take that in to account. CLI's test runner points those out now, but those issues were already there. Those tests have never actually tested anything. We felt that fixing them all was out of the scope for this PR, and should be tackled in multiple small PRs

---

#### Update 2020-03-27
The config system has been overhauled. 

Environment files no longer need to be on git. The system also no longer crashes when there's no custom dev or prod config. We support environment variables, and we also added support for .env files. 

We investigated hosting the environment on an endpoint so it could be updated without the need for a rebuild, but that didn't turn out to be feasible in the time we had.

For more info see [Readme.md](https://github.com/DSpace/dspace-angular/blob/f61f92e1ad1c7b5fc0a16547a8297f71437d47e1/README.md#configuring)

### CODE REVIEW

**Comment by atarix83** [comment]
thanks @artlowel for this important PR

I have to premise that I've only had a quick overview but I'll write you first questions/doubts right now.
There is a specific reason for removing `deploy/undeploy` scripts? Before we use pm2 as node process manager, why do not make it coexist with CLI? 

> The preboot property does not exist anymore, in cli the server side is built for production builds only.

This could impact negatively on development: watching the behaviour of the application when works with SSR is very useful sometimes. However if there are no other possible solutions, we'll adapt to it and we'll need to be more careful that SSR works properly

**Comment by artlowel** [comment]
>There is a specific reason for removing deploy/undeploy scripts? Before we use pm2 as node process manager, why do not make it coexist with CLI?

We felt that there was no reason for pm2 to be a direct dependency of dspace-angular. We have servers using pm2 and others using forever. We felt that that choice shouldn't be made for you by dspace. 

>>The preboot property does not exist anymore, in cli the server side is built for production builds only.
>
>This could impact negatively on development: watching the behaviour of the application when works with SSR is very useful sometimes. However if there are no other possible solutions, we'll adapt to it and we'll need to be more careful that SSR works properly

That wasn't a choice we made. It is the way it works for universal development on cli at the moment. however [the universal team is working on improving this](https://github.com/angular/universal/issues/1022), in the near future it will likely no longer be necessary. For now though the benefit is much faster rebuilds when you make a change during development.

**Comment by atarix83** [comment]
> That wasn't a choice we made. It is the way it works for universal development on cli at the moment. however the universal team is working on improving this, in the near future it will likely no longer be necessary. For now though the benefit is much faster rebuilds when you make a change during development.

This is a good news :+1: 

> We felt that there was no reason for pm2 to be a direct dependency of dspace-angular. We have servers using pm2 and others using forever. We felt that that choice shouldn't be made for you by dspace.

We don't see it as a dspace constraint, but I see it as added value. Out of the box dspace already has scripts to deploy/undeploy quickly and easily. Our opinion is not to remove it especially because it is an already existing option.

**Comment by mwoodiupui** [comment]
On Thu, Mar 19, 2020 at 03:11:52AM -0700, Art Lowel wrote:
> >There is a specific reason for removing deploy/undeploy scripts? Before we use pm2 as node process manager, why do not make it coexist with CLI?
> 
> We felt that there was no reason for pm2 to be a direct dependency of dspace-angular. We have servers using pm2 and others using forever. We felt that that choice shouldn't be made for you by dspace. 

Thank you -- I agree.

-- 
Mark H. Wood
Lead Technology Analyst

University Library
Indiana University - Purdue University Indianapolis
755 W. Michigan Street
Indianapolis, IN 46202
317-274-0749
www.ulib.iupui.edu

File: package.json [hunk 1]
```diff
@@ -24,105 +38,58 @@
...
-    "test:headless": "karma start --single-run --browsers ChromeHeadless",
-    "test:watch": "karma start --no-single-run --auto-watch",
-    "webdriver:start": "node node_modules/protractor/bin/webdriver-manager start --seleniumPort 4444",
-    "webdriver:update": "node node_modules/protractor/bin/webdriver-manager update --standalone --gecko false",
-    "lint": "tslint \"src/**/*.ts\" && tslint \"e2e/**/*.ts\"",
-    "docs": "typedoc --options typedoc.json ./src/",
-    "coverage": "http-server -c-1 -o -p 9875 ./coverage",
-    "postinstall": "yarn run patch-protractor",
-    "patch-protractor": "ncp node_modules/webdriver-manager node_modules/protractor/node_modules/webdriver-manager",
-    "sync-i18n": "node ./scripts/sync-i18n-files.js"
```
Comment (atarix83) [inline package.json:75]: sync-i18n is no longer available?

**Comment by LotteHofstede** [comment]
@tdonohue
Thanks for offering to help me with the e2e issues I'm experiencing with Travis (locally it seems to work), here's a list of things I tried to solve this: 


- Added more `then`s to e2e test https://github.com/DSpace/dspace-angular/pull/625/commits/45beb3f93fab59f26605c0aff08126840c7013be
- Added a method that waits until there's no more "Loading..." visible on the page https://github.com/DSpace/dspace-angular/pull/625/commits/97953912c55f5cb30b9cfe564fd09aaab64a8590 
- Increased timeouts to 3 minutes (180000) https://github.com/DSpace/dspace-angular/pull/625/commits/51dd1b1b2b787b2db740a2551cba44e2f10e879a
- Switched everything back to port 3000 https://github.com/DSpace/dspace-angular/pull/625/commits/172d3891f05e0995acb4ea6ccc72310ed55c7f76
- Used the 4Science REST server again, instead of the docker instance https://github.com/DSpace/dspace-angular/pull/625/commits/ad4b7fa254c3e8114aac67012a0bd371f60fc11f
- Ran just the e2e (without other tests https://github.com/DSpace/dspace-angular/pull/625/commits/2e367cb87c26c14f54439927a70d1e32d071e15d)

All these changes did not seem to make a difference.

I also tried setting up a Travis container in Docker, in hope of debugging the issue there, but all the tests succeeded there as well. 
However, I am fairly certain that my Docker container setup is not exactly the same as that of the 'real' Travis, because locally I had to add "--no-sandbox" to the Chrome settings for both the Karma and Protractor tests to get it to run the tests at all.
I don't have much experience with Docker, so I might have done something wrong.

I used this tutorial to run Travis in Docker: https://jonlabelle.com/snippets/view/markdown/run-travis-ci-builds-locally-with-docker
Together with this image: travisci/ci-ubuntu-1804:packer-1582574251-a9b1ae58

I could not find much about it online, all I could find was this StackOverflow post (that I'm not sure is even related): https://stackoverflow.com/questions/54966792/dockerized-angular-e2e-tests-pass-on-local-machine-but-timeout-in-travis-ci/54992080

I hope this helps! If you have any other questions, let me know.

**Comment by tdonohue** [comment]
@LotteHofstede : Thanks for the info. I've dug a bit yesterday and today, and I'm fairly certain the problem is *not* with Docker.  The Docker containers are starting up fine, and if I add this to the `.travis.yml`, I see an immediate response from the Docker-based REST API:
```
before_script:
  - curl http://localhost:8080/server/
```

This runs *before* protractor kicks anything off, so I can verify Docker has started by then.  It seems to me like something is not running properly *within Protractor or the Chrome headless browser*.  I've yet to narrow down exactly what the issue is, but I'm almost certain it's not Docker that is the problem here.

If I discover more, I'll let you know. But, I wanted to let you know I've been working on this a bit behind the scenes (I'm just using Travis with my personal acct to run tests to see if I can figure out what is going on, rather than adding to this PR.)

**Comment by tdonohue** [comment]
@LotteHofstede :

I've narrowed down the issue further.  It's *not* Docker, but the core problem is that the `<ds-loading>` tag never goes away (for any of the e2e tests).  **So, the "loading..." message always appears for every e2e test**.  The only e2e tests that currently succeed are those that depend on parts of the page which do *not* end up appearing where the `<ds-loading>` tag starts at.  You can see this behavior by adding the equivalent of a `page.waitUntilNotLoading();` inside every single e2e test. When that is done, they all fail with a timeout, as the "loading" message never disappears from the app.

This bit of debug code can also be used to print out the full HTML that protractor sees in a given test. Every time I do this in an e2e test it prints out the HTML page with the "loading" message (beware it prints out a lot of CSS as well, so you have to jump to the bottom to find the HTML portion of the page): 
```
    // This block of code will print the source of page (for debugging test issues)
    browser.getPageSource().then(function (res) {
        console.log("Page Source code is = " + res);
    });
```

So, from what I can tell, the issue seems to be something in how Protractor is loading our Angular CLI app.  I'm not sure why that would be, but it seems to have nothing to do with the REST API availability (Docker is working). _My best guess is it must be something in how Protractor is configured now that we've updated to Angular CLI._  But, I'm not sure exactly where the problem is quite yet.

File: scripts/set-env.ts [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+
+// allow to override a few important options by environment variables
+function createServerConfig(host?: string,  port?: string, nameSpace?: string, ssl?: string): ServerConfig {
+  const result = {} as any;
+  if (hasValue(host)) {
+    result.host = host;
+  }
+
+  if (hasValue(nameSpace)) {
+    result.host = nameSpace;
```
Comment (tdonohue) [inline scripts/set-env.ts:75]: Here's the small but **annoying** bug in our e2e tests.  This needs to be:
`result.nameSpace = nameSpace;`

This was causing all the issues, as we depend on the `DSPACE_REST_*` env variables for our tests.  Obviously, this current setting caused an invalid REST API URL to be used by the Angular app.

**Comment by LotteHofstede** [comment]
@tdonohue Thank you so much!! That makes a lot of sense, actually!
I'll reverse the unnecessary changes (but in a new commit so the comments in this PR don't become messed up) and I'll apply your recommended changes to the travis.yml, and of course I'll fix the issue in set-env.ts.

Thanks again, you saved me a lot of time!

**Comment by tdonohue** [comment]
@LotteHofstede: Sounds good.  It looks like the latest commit still had failures (though the failures are a bit different this time around).  It *is* possible the `.travis.yml` changes caused unexpected new issues. If so, we can move that to a separate follow-up PR.  I'll leave it up to you though :)  Let me know if you need any further help, otherwise I'll wait for re-review / re-test once you are ready.

**Comment by LotteHofstede** [comment]
@tdonohue The tests fail because of the fix, woops... There is a unit test that relies on a `nameSpace` value configured in the environment.test.ts file, but it's now overwritten by Travis' env-variable value for the `nameSpace` (that does get set now I've applied your fix).
I'm working on a solution that makes sure unit tests don't use the set-env.ts script at all, because they should be independent from the environment they run in.

**Comment by tdonohue** [comment]
@atarix83 : Can you re-review this PR?  It looks to me like your previous requests/questions have all been resolved, but it'd be good to get your approval.

**Comment by LotteHofstede** [comment]
@atarix83 The script is now a typescript file instead of a javascript one, because it relies on other typescript files.
To run the script from your console* you'll need use ts-node like this:
`ts-node  --project ../../../tsconfig.ts-node.json ../../../scripts/sync-i18n-files.ts -s en.json5 -i -t cz.json5`

*) Note that the wiki page runs the script from `resources/i18n`, but the i18n folder is now located at `src/assets/i18n`, so I executed it in that location.

**Comment by atarix83** [comment]
thanks @LotteHofstede it works, maybe is better to update the wiki after this PR is merged

**Comment by tdonohue** [comment]
This is at +2 and ready to be merged.  @artlowel will merge it tomorrow as merger will likely require immediate updates to the demo Angular UI (once this gets deployed).

After merger, as noted by @atarix83 , we also should update the documentation on the `sync-i18n-files` script on this wiki page: https://wiki.lyrasis.org/pages/viewpage.action?pageId=117735441#DSpace7TranslationInternationalization(i18n)Localization(l10n)-Startinganewlanguagetranslationfromscratch

**Pull request merged**